### PR TITLE
Backbone Gap/Stagger AdaLN: Thread Tandem Geometry Into All Blocks

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The gap and stagger scalars (tandem foil spacing and offset) are currently only available to the LAST TransolverBlock via `adaln_output` conditioning with (Re, AoA). **None of the backbone TransolverBlocks know they are processing a tandem configuration or what the gap/stagger is.** The slice assignment (which mesh nodes get grouped into which physics slice) is the core computational primitive — and for tandem foils, the inter-foil gap changes the pressure coupling length scale. Nodes near the gap should attend to different neighbors depending on gap width.

The infrastructure for conditioning ALL blocks already exists via `adaln_all_blocks=True`, which creates an `adaln_net` in every TransolverBlock that modulates both pre-attention and pre-MLP LayerNorm via AdaLN-Zero (scale+bias, zero-initialized so training starts as the unconditioned baseline). Currently this path uses only (Re, AoA) as condition. By changing the condition to include gap and stagger, every block's attention routing becomes tandem-geometry-aware from layer 1.

**Why this is different from GSB (PR #2130):** GSB injects gap/stagger into the `spatial_bias` MLP that feeds the attention scores. This gives the attention *keys* access to geometry, but not the *queries* or *values*, and only through a single scalar bias per head. AdaLN-All conditions the ENTIRE residual stream (both attention and FFN sublayers) by modulating the LayerNorm, which affects all dimensions of all attention heads simultaneously. The two mechanisms are complementary — GSB does per-head routing adjustment, AdaLN does global representation modulation.

**This is a bold structural change** — it adds new trainable parameters to all 3 backbone blocks and fundamentally changes how information flows through the model for tandem vs single-foil samples.

### Step 1: Enable `adaln_all_blocks` with gap/stagger condition

The code at lines ~857-868 in `Transolver.forward` builds `block_condition`. Currently with `adaln_4cond=True` it uses `(Re, AoA, gap_feat, surf_frac)`. We want `(Re, AoA, gap, stagger)` — using the same gap/stagger features that GSB and aft-foil SRF use.

**Modify the condition construction** (search for `if self.adaln_4cond:` in `Transolver.forward`):

```python
if self.adaln_4cond:
    # CHANGE: use gap and stagger (indices 22:24) instead of gap_feat and surf_frac
    gap_stagger = x[:, 0, 22:24]  # [B, 2] — gap and stagger from raw input
    block_condition = torch.cat([cond_2, gap_stagger], dim=-1)  # [B, 4] = (Re, AoA, gap, stagger)
```

This replaces the old `gap_feat` (index 21) and `surf_frac` (computed boolean) with the actual gap/stagger scalars that have proven informative throughout Phase 6.

**For single-foil samples**, gap=0 and stagger=0 in the data, so the condition naturally becomes (Re, AoA, 0, 0) — no special masking needed.

### Step 2: No changes to TransolverBlock needed

The `adaln_net` in each block already handles 4-dim input when `adaln_cond_dim=4`. The `adaln_all` codepath (lines 403-409) modulates `ln_1` and `ln_2` via the standard AdaLN-Zero pattern. Zero-init ensures the model starts as if unconditioned.

### Step 3: Run 2 configs

**Config A: adaln_all + adaln_4cond with gap/stagger** (the main hypothesis)
```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent frieren \
  --wandb_name "frieren/backbone-gs-adaln-s42" --wandb_group phase6/backbone-gs-adaln \
  --adaln_all_blocks --adaln_4cond --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73: same flags, --seed 73, wandb_name "frieren/backbone-gs-adaln-s73"
```

**Config B (control): adaln_all WITHOUT 4cond — just Re/AoA in all blocks**
```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent frieren \
  --wandb_name "frieren/backbone-adaln-reao-s42" --wandb_group phase6/backbone-gs-adaln \
  --adaln_all_blocks --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73: same flags, --seed 73, wandb_name "frieren/backbone-adaln-reao-s73"
```

Config B (Re/AoA only) isolates whether conditioning all blocks helps at all, vs Config A which tests the gap/stagger addition specifically.

### What to report

| Config | Seed | p_in | p_oodc | p_tan | p_re | W&B |
|--------|------|------|--------|-------|------|-----|
| adaln_all + gs 4cond | 42 | | | | | |
| adaln_all + gs 4cond | 73 | | | | | |
| **gs 4cond avg** | — | | | | | |
| adaln_all Re/AoA only | 42 | | | | | |
| adaln_all Re/AoA only | 73 | | | | | |
| **Re/AoA only avg** | — | | | | | |
| **Baseline** | — | 13.05 | 7.70 | 28.60 | 6.55 | d7l91p0x, j9btfx09 |

Also log VRAM usage for each config (the adaln_net adds parameters to all 3 blocks — expect a small increase).

## Baseline

| Metric | Target |
|--------|--------|
| p_in   | < 13.05 |
| p_oodc | < 7.70  |
| **p_tan** | **< 28.60** |
| p_re   | < 6.55  |

W&B baseline: d7l91p0x (s42), j9btfx09 (s73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Summary

**Both configs are significantly worse than baseline on p_tan.** Config A (adaln_all + gap/stagger 4cond): p_tan +6.8%. Config B (adaln_all + Re/AoA only): p_tan +5.9%. The backbone AdaLN mechanism hurts tandem transfer regardless of the condition content. p_oodc and p_re are flat/marginal; p_in is mixed.

### Metrics

| Config | Seed | p_in | p_oodc | p_tan | p_re | W&B |
|--------|------|------|--------|-------|------|-----|
| adaln_all + gs 4cond | 42 | 13.1 | 7.6 | 30.3 | 6.5 | `8td4pkvl` |
| adaln_all + gs 4cond | 73 | 13.3 | 7.8 | 30.8 | 6.6 | `qwbhkksf` |
| **gs 4cond avg** | — | **13.2** | **7.7** | **30.55** | **6.55** | |
| adaln_all Re/AoA only | 42 | 14.3 | 8.0 | 30.4 | 6.6 | `z0qjvsiv` |
| adaln_all Re/AoA only | 73 | 13.4 | 7.6 | 30.2 | 6.5 | `m5y5lfy2` |
| **Re/AoA only avg** | — | **13.85** | **7.8** | **30.3** | **6.55** | |
| **Baseline** | — | 13.05 | 7.70 | 28.60 | 6.55 | d7l91p0x, j9btfx09 |

| Config | Δ p_in | Δ p_oodc | Δ p_tan | Δ p_re |
|--------|--------|----------|---------|--------|
| A (gs 4cond) | +1.1% | 0% | **+6.8%** | 0% |
| B (Re/AoA) | +6.1% | +1.3% | **+5.9%** | 0% |

### VRAM Usage
- Config A (adaln_all + 4cond): 49.3-50.2 GB (vs ~44-46 GB baseline, +4-6 GB)
- Config B (adaln_all, 2-dim): 49.8-50.2 GB (same increase)
- Epochs: 142-144 in 180 min (vs 150-152 baseline) — ~83-85s/epoch vs ~70-72s baseline (+18% slower)

### Commands

```bash
# Config A: adaln_all + adaln_4cond (gap/stagger condition)
cd cfd_tandemfoil && python train.py --agent frieren \
  --wandb_name "frieren/backbone-gs-adaln-s42" --wandb_group phase6/backbone-gs-adaln \
  --adaln_all_blocks --adaln_4cond --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Config B: adaln_all without adaln_4cond (Re/AoA only)
# Same but remove --adaln_4cond flag
```

### Code change

Modified `adaln_4cond` condition from `[Re, AoA, gap_feat(idx 21), surf_frac(computed)]` to `[Re, AoA, gap(idx 22), stagger(idx 23)]` — using the same gap/stagger scalars as GSB. Single-foil samples get (Re, AoA, 0, 0) naturally.

### What happened

**The adaln_all_blocks mechanism itself is the problem, not the condition content.** Both configs regress equally on p_tan (~+6%), despite Config A having geometry-aware conditioning and Config B only having Re/AoA. This suggests:

1. **Modulating ALL backbone LayerNorms interferes with the already-optimized attention routing.** The baseline uses GSB (Gap/Stagger Spatial Bias) which affects attention scores directly — a light-touch mechanism. AdaLN modulates the full residual stream via LayerNorm scale/bias, which is a much heavier intervention that likely disrupts the fine-tuned balance between single-foil and tandem representations.

2. **The zero-init doesn't save it.** While AdaLN-Zero starts as identity, within ~140 epochs the learned modulations drift enough to hurt tandem transfer. The model may learn to over-specialize the backbone for single-foil cases (which are majority in training) at the cost of tandem generalization.

3. **Fewer epochs due to increased compute.** 142-144 epochs vs 150-152 baseline means the model didn't train as long, though the gap is small (~5%) and unlikely to explain the full regression.

4. **Config A (gap/stagger) is slightly better on p_oodc (7.7 = baseline) but worse on p_tan (30.55 vs 30.3).** The gap/stagger information helps OOD conditioning but doesn't rescue tandem transfer — the fundamental adaln_all problem dominates.

### Suggested follow-ups

1. **Condition ONLY the last block** instead of all blocks — `adaln_output` already does this for (Re, AoA), try extending it to also include gap/stagger in the output block conditioning only.
2. **Lighter-weight conditioning:** Instead of full AdaLN (scale+bias on both ln_1 and ln_2), try FiLM-conditioning on just one LayerNorm per block, or only on the FFN path.
3. **Consider this a negative result for backbone-wide conditioning.** The current architecture already has sufficient tandem awareness via GSB + tandem_ramp + aft_foil_srf. Adding more global conditioning adds parameters and complexity without benefit.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2164 because ADVISOR: Closing — both configurations regress p_tan heavily.

**Config A (adaln_all + gap/stagger 4-cond):** avg p_tan=30.55 (+6.8% vs 28.60 baseline)
**Config B (adaln_all + Re/AoA 2-cond):** avg p_tan=30.3 (+5.9% vs baseline)

Key finding: modulating ALL backbone LayerNorms via AdaLN-Zero interferes with the already-optimized attention routing and slice assignments in the Transolver blocks. The backbone learned its routing patterns during 160 epochs of training — adding AdaLN conditioning on top disrupts this. The GSB mechanism (which works at the spatial-bias level, not the normalization level) is the right abstraction for injecting geometry info.

Also note VRAM jumped to ~50GB and effective epochs dropped to 142-144, so the model was also under-trained relative to baseline.

Dead end: backbone-wide AdaLN conditioning. GSB remains the correct mechanism for tandem geometry injection.

Thanks frieren — this was a high-priority hypothesis and the negative result is very informative..
